content/posts/numpy/numpy-rng/index.md

```diff
@@ -0,0 +1,136 @@
+
+In short:
+* Instead of using `np.random.seed`, which reseeds the already created global NumPy RNG and then using `np.random.*` functions you should create a new RNG.
+* You should create one RNG at the beginning of your script (with a seed if you want reproducibility) and use this RNG in the rest of your script.
+
+To create a new RNG you can use the `default_rng` function as illustrated in the [introduction of the random module documentation](https://numpy.org/doc/stable/reference/random/index.html):
+
+```python
+import numpy as np
+
+rng = np.random.default_rng()
+rng.random()
+```
+
+If you want to use a seed for reproducibility, [Numpy documentation](https://numpy.org/doc/stable/reference/random/index.html#quick-start) recommends using a very large and unique number to have a different seed than anyone else. Another reason is that using a small number for the seed could lead to a bad initialization of the RNG, although `default_rng` now takes care of creating a good initialization from a small number. To obtain a good unique seed you can rely on the [secrets module](https://docs.python.org/3/library/secrets.html). The following code will return a 128 bit number that you can then use as a seed:
+
+```python
+import secrets
+
+secrets.randbits(128)
+```
+
+When running this code I get `65647437836358831880808032086803839626` for the number to use as my seed. This number is randomly generated so you need to copy paste the value that is returned by `secrets.randbits(128)` otherwise you will have a different seed each time you run your code and thus break reproducibility:
+
+```python
+import numpy as np
+
+seed = 65647437836358831880808032086803839626
+rng = np.random.default_rng(seed)
+rng.random()
+```
+
+The reason for seeding your RNG only once (and passing that RNG around) is that you can lose on the randomness and the independence of the generated random numbers by reseeding the RNG multiple times. Furthermore obtaining a good seed can be time consuming. Once you have a good seed to instantiate your generator you might as well use it. With a good RNG such as the one returned by `default_rng` you will be ensured good randomness (and independence) of the generated numbers. It might be more dangerous to use different seeds: how do you know that the streams of random numbers obtained with two different seeds are not correlated, or I should say less independent than the ones created from the same seed? That being said, [as explained by Robert Kern](https://github.com/numpy/numpy/issues/15322#issuecomment-573890207), with the RNGs and seeding strategies introduced in NumPy 1.17, it could be considered safe enough to recreate new RNGs from the system entropy, e.g. using `default_rng(None)` multiple times. However as explained later be careful when running jobs in parallel and relying on `default_rng(None)`.
```

Comment:
```suggestion
The reason for seeding your RNG only once (and passing that RNG around) is that you can lose on the randomness and the independence of the generated random numbers by reseeding the RNG multiple times. Furthermore obtaining a good seed can be time consuming. Once you have a good seed to instantiate your generator, you might as well use it. With a good RNG such as the one returned by `default_rng` you will be ensured good randomness (and independence) of the generated numbers. It might be more dangerous to use generators derived from different seeds: how do you know that the streams of random numbers obtained from those are not correlated or, I should say, less independent than the ones created from the same seed? [As explained by Robert Kern](https://github.com/numpy/numpy/issues/15322#issuecomment-573890207), with the RNGs and seeding strategies introduced in NumPy 1.17, it is considered fairly safe to create RNGs using system entropy, i.e. using `default_rng(None)` multiple times. However as explained later be careful when running jobs in parallel and relying on `default_rng(None)`.
```

Reply:
It's a bit unclear how the sentences on `default_rng(None)` relate to seeding. Above, you say not to use `None` as the input to `default_rng`, and now that is being suggested, so would be worth clarifying when that's a good idea.

Reply:
yes good point again, what's not clear in my text is that `default_rng(None)` will still need to obtain a seed from somewhere (e.g. system entropy)